.github/setup-node/action.yml

```diff
@@ -0,0 +1,58 @@
+name: 'Setup Node'
+description: 'Setup node and install dependencies with caching support'
+inputs:
+    node-version:
+        description: 'Optional node version to pass to setup-node'
+        required: false
+        type: string
+
+    should-rebuild:
+        description: 'Whether to rebuild the dependencies after --ignore-scripts'
+        default: false
+        required: false
+        type: boolean
+
+runs:
+    using: 'composite'
+    steps:
+        - name: Use desired version of NodeJS from input
+          if: ${{ inputs.node-version != '' }}
+          uses: actions/setup-node@v3
+          with:
+              node-version: ${{ inputs.node-version }}
+              cache: npm
+
+        - name: Use desired version of NodeJS from .nvmrc
+          if: ${{ inputs.node-version == '' }}
+          uses: actions/setup-node@v3
+          with:
+              node-version-file: '.nvmrc'
+              cache: npm
+
+        - name: Get the node version
+          id: node-version
+          run: echo "NODE_VERSION=$(node -v)" >> $GITHUB_OUTPUT
+          shell: bash
+
+        - name: Cache node_modules
+          id: cache-node_modules
+          uses: actions/cache@v3
+          with:
+              path: '**/node_modules'
+              key: node_modules-${{ runner.os }}-${{ steps.node-version.outputs.NODE_VERSION }}-${{ hashFiles('package-lock.json') }}
+
+        - name: Npm install
+          if: ${{ steps.cache-node_modules.outputs.cache-hit != 'true' }}
+          run: npm ci --ignore-scripts
+          shell: bash
+
+        - name: Rebuild dependencies
+          # Comparing against strings is necessary because of a known bug in composite actions.
+          # See https://github.com/actions/runner/issues/1483#issuecomment-1279933184
+          if: ${{ inputs.should-rebuild == 'true' }}
+          run: npm rebuild
+          shell: bash
+
+        - name: Post-install
+          run: npm run postinstall
+          shell: bash
```

Comment:
quick note here: we appear to have prior art in creating a new package for this: see `packages/report-flaky-tests` and `packages/project-management-automation`

I don't have a good feel for preferring one over the other, but it does appear that our existing custom GitHub Actions are created as new packages.

Reply:
It's not a new action though, it's just a composite action. Which simply means that it's an abstraction that prevents us from repeating ourselves. We can move it to the `packages` folder but I think that would be overkill at this point.

Reply:
whatever we call it, I just wanted to note that we have precedence. maybe @youknowriad would have thoughts.

Reply:
Personally, I am against having any GitHub Action specific code outside of the `.github/workflows` directory. The only times we should make exceptions are when there is a reason to run a script.

A good example is [adding a first time contributor label](https://github.com/WordPress/gutenberg/blob/trunk/packages/project-management-automation/lib/tasks/first-time-contributor-label/index.js). This script would be _very_ awkward to maintain within YAML files.

This change takes advantage of a type of workflow offered in GHA to group a set of shared steps. One or several one line commands grouped together is probably overkill for a new package. So in my opinion, it feels right to include it here rather than within a package (which also has other implications like supporting the package because it was released publicly).